File: app/src/main/java/com/github/overpass/gather/auth/register/signup/SignUpUseCase.java
package com.github.overpass.gather.auth.register.signup;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.github.overpass.gather.auth.BaseValidator;

public class SignUpUseCase {

    private final SignUpRepo signUpRepo;
    private final BaseValidator validator;

    SignUpUseCase(SignUpRepo signUpRepo, BaseValidator validator) {
        this.signUpRepo = signUpRepo;
        this.validator = validator;
    }

    LiveData<SignUpStatus> signUp(String email, String password) {
        if (!validator.isEmailValid(email)) {
            MutableLiveData<SignUpStatus> signUpStatus = new MutableLiveData<>();
            signUpStatus.setValue(new SignUpStatus.InvalidEmail("Invalid Email"));
            return signUpStatus;
        }
        if (!validator.isPasswordValid(password)) {
            MutableLiveData<SignUpStatus> signUpStatus = new MutableLiveData<>();
            signUpStatus.setValue(new SignUpStatus.InvalidPassword("Invalid Password"));
            return signUpStatus;
        }
        return signUpRepo.signUp(email, password);
    }

}

File: app/src/main/java/by/overpass/gather/model/usecase/register/SignUpUseCase.kt
package by.overpass.gather.model.usecase.register

import by.overpass.gather.di.EMAIL_VALIDATOR
import by.overpass.gather.di.PASSWORD_VALIDATOR
import by.overpass.gather.commons.abstractions.Result
import by.overpass.gather.model.data.validator.Validator
import by.overpass.gather.model.repo.register.SignUpRepo
import by.overpass.gather.ui.auth.register.signup.SignUpStatus
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.flow.map
import javax.inject.Inject
import javax.inject.Named

class SignUpUseCase @Inject constructor(
        private val signUpRepo: SignUpRepo,
        @Named(EMAIL_VALIDATOR) private val emailValidator: Validator<String>,
        @Named(PASSWORD_VALIDATOR) private val passwordValidator: Validator<String>
) {

    @ExperimentalCoroutinesApi
    suspend fun signUp2(email: String, password: String): Flow<SignUpStatus> {
        return if (!emailValidator.isValid(email)) {
            flowOf(SignUpStatus.InvalidEmail("Invalid Email"))
        } else if (!passwordValidator.isValid(password)) {
            flowOf(SignUpStatus.InvalidPassword("Invalid Password"))
        } else {
            signUpRepo.signUp(email, password)
                    .map {
                        when (it) {
                            is Result.Loading -> SignUpStatus.Progress
                            is Result.Success -> SignUpStatus.Success
                            is Result.Error -> SignUpStatus.Error(it.exception)
                        }
                    }
        }
    }
}
